vsintegration/src/FSharp.Editor/CodeFixes/CodeFixHelpers.fs

```diff
@@ -106,3 +108,57 @@ module internal CodeFixExtensions =
                 let! sourceText = ctx.GetSourceTextAsync()
                 return RoslynHelpers.TextSpanToFSharpRange(ctx.Document.FilePath, ctx.Span, sourceText)
             }
+
+// This cannot be an extension on the code fix context
+// because the underlying GetFixAllProvider method doesn't take the context in.
+#nowarn "3511" // state machine not statically compilable
+
+[<AutoOpen>]
+module IFSharpCodeFixProviderExtensions =
+    type IFSharpCodeFixProvider with
+
+        member provider.RegisterFsharpFixAll() =
+            FixAllProvider.Create(fun fixAllCtx doc allDiagnostics ->
+                cancellableTask {
+                    let sw = Stopwatch.StartNew()
+
+                    let! token = CancellableTask.getCurrentCancellationToken ()
+                    let! sourceText = doc.GetTextAsync token
+
+                    // this is not used anywhere, it's just needed to create the context
+                    let action =
+                        Action<CodeActions.CodeAction, ImmutableArray<Diagnostic>>(fun _ _ -> ())
+
+                    let! codeFixOpts =
+                        allDiagnostics
+                        // The distiction is to avoid collisions of compiler and analyzer diags
+                        // See: https://github.com/dotnet/fsharp/issues/15620
+                        |> Seq.distinctBy (fun d -> d.Id, d.Location)
+                        |> Seq.map (fun diag -> CodeFixContext(doc, diag, action, token))
+                        |> Seq.map (fun context -> provider.GetCodeFixIfAppliesAsync context)
+                        |> Seq.map (fun task -> task token)
```

Comment:
I don't think that this is wise, at least without any proofs on large documents.
Before there was 1 backgroundTask for entire document, with this change there is a started task for each diagnostics in a file, all running in parallel.

Reply:
Hm I cannot imagine a scenario where this could be any real-world problem. Also, in the vast majority of cases - as in, a user is writing code and makes a mistake - that would effectively just produce a cancellable task instead of a background task, which should be more efficient.

Regardless, I haven't noticed any weird behavior when trying this out on our solution and inserting random errors here and there. We can optimize this later, if anybody reports any problems or when we start measuring things properly here. 

Probably once we are done reworking the code fixes, we can create some project full of random diags and create some way of applying all the code fixes possible and measuring that. But that's a separate thing and not a very trivial one.

Reply:
> a user is writing code and makes a mistake - that would effectively just produce a cancellable task instead of a background task, which should be more efficient.

Is this code getting invoked for document every time when user makes a mistake? In this case we probably shouldn't be invoking N many tasks in parallel. This will make VS unresponsive very quick, and scheduler won't be scheduling important things instead of codefixes.

Reply:
No sorry I was just editing the message to fix this :facepalm: 

This is for bulk code fixes only. So a scenario would be - e.g. a user is cleaning up the code after refactoring and removes a few unused bindings in the function. Or fixes the whole pattern matching block for the redundant bindings. Something like this.

Reply:
Oh, ok, it's an explicit user action then. I'm personally fine with it then. We do the same in find all references, and I am doing the same for refactorings.

Reply:
Fixing "random" or "when typing" errors isn't my worry.
It is rather about activating a codefix on a big codebase for the first time ever - e.g. because it is a new feature, or the team did not know about this option before, etc.

I think it would be best and clearest to add a basic test at scale= proof of doing really a big bulk codefix.
e.g. swapping `.[]` and `[]` indexer access, something which can easily have over hundreds of occurences.

Reply:
No I agree this is important - again I just suggest we do it separately as we need to decide where to put it and what use cases we'll have there. The same goes for renaming, the same goes for finding references, and so on. IMO It would be best to start adding this once we come up with at least some performance guidelines to align with.

As a mini stress test, I tried applying the code fix in the current shape for 100 occurrences in 5 docs, this works smoothly (honestly smoother than I expected :D):

https://github.com/dotnet/fsharp/assets/5451366/26bd3f2a-ffb7-454e-b43d-116b22343c86

There is no noticeable difference for the main version of this code.